lib/AST/RequirementMachine/RewriteSystem.cpp

```diff
@@ -705,10 +718,99 @@ void RewriteSystem::verifyRewriteRules(ValidityPolicy policy) const {
 #undef ASSERT_RULE
 }
 
+std::vector<RequirementError>
+RewriteSystem::getRedundantRequirements() {
+  // The resulting set of redundant requirements.
+  std::vector<RequirementError> redundantRequirements;
+
+  // Map redundant rule IDs to their rewrite path for easy access
+  // in the `isRedundantRule` lambda.
+  llvm::SmallDenseMap<unsigned, RewritePath> redundantRules;
+  for (auto &pair : RedundantRules)
+    redundantRules[pair.first] = pair.second;
+
+  // Collect all rule IDs for each unique requirement ID.
+  llvm::SmallDenseMap<unsigned, llvm::SmallDenseSet<unsigned, 2>>
+      rulesPerRequirement;
+
+  // Collect non-explicit requirements that are not redundant.
+  llvm::SmallDenseSet<unsigned, 2> impliedRequirements;
+
+  for (unsigned ruleID : indices(getRules())) {
+    auto &rule = getRules()[ruleID];
+
+    if (!rule.isExplicit() && !rule.isPermanent() && !rule.isRedundant() &&
+        isInMinimizationDomain(rule.getLHS().getRootProtocol()))
+      impliedRequirements.insert(ruleID);
+
+    if (!rule.isExplicit())
+      continue;
+
+    auto requirementID = rule.getRequirementID();
+    if (!requirementID.hasValue())
+      continue;
+
+    rulesPerRequirement[*requirementID].insert(ruleID);
+  }
+
+  auto isRedundantRule = [&](unsigned ruleID) {
+    auto &rule = getRules()[ruleID];
+
+    // If this rule is replaced using a non-explicit,
+    // non-redundant rule, it's not redundant.
+    auto rewritePath = redundantRules[ruleID];
```

Comment:
Should RewriteSystem::RedundantRules be a MapVector then so that we can do lookups while preserving insertion order?